server/athenian/api/controllers/miners/github/pull_request.py

```diff
@@ -0,0 +1,207 @@
+from dataclasses import dataclass
+from datetime import datetime
+from enum import Enum
+from typing import Generic, Optional, Sequence, TypeVar, Union
+
+import databases
+import pandas as pd
+from sqlalchemy import select, sql
+
+from athenian.api.async_read_sql_query import read_sql_query
+from athenian.api.models.metadata.github import PullRequest, PullRequestComment, PullRequestReview
+
+
+class PullRequestMiner:
+    """Load all the information related to Pull Requests from the metadata DB. Iterate over it \
+    with individual PR tuples."""
+
+    def __init__(self, prs: pd.DataFrame, reviews: pd.DataFrame, review_comments: pd.DataFrame):
+        """Initialize a new instance of `PullRequestMiner`."""
+        self._prs = prs
+        self._reviews = reviews
+        self._review_comments = review_comments
+
+    @classmethod
+    async def mine(cls, time_from: datetime, time_to: datetime, repositories: Sequence[str],
+                   developers: Sequence[str], db: databases.Database) -> "PullRequestMiner":
+        """
+        Create a new `PullRequestMiner` from the metadata DB according to the specified filters.
+
+        :param time_from: Fetch PRs created starting from this date.
+        :param time_to: Fetch PRs created ending with this date.
+        :param repositories: PRs must belong to these repositories (prefix excluded).
+        :param developers: PRs must be authored by these user IDs. An empty list means everybody.
+        :param db: Metadata db instance.
+        """
+        repos_parsed = [r.split("/") for r in repositories]
+        repo_orgs = [r[0] for r in repos_parsed]
+        repo_names = [r[1] for r in repos_parsed]
+        filters = [
+            PullRequest.created_at >= time_from,
+            PullRequest.created_at < time_to,
+            PullRequest.repository_name.in_(repo_names),
+            PullRequest.repository_owner.in_(repo_orgs),  # FIXME(vmarkovtsev): ENG-101 makes this correct # noqa
+        ]
+        if len(developers) > 0:
+            filters.append(PullRequest.user_login.in_(developers))
+        prs = await read_sql_query(select([PullRequest]).where(sql.and_(*filters)), db)
+        if len(prs) == 0:
+            return cls(prs, prs, prs)
+        numbers = prs["number"]
+        reviews = await read_sql_query(select([PullRequestReview]).where(
+            PullRequestReview.pull_request_number.in_(numbers)), db)
+        review_comments = await read_sql_query(select([PullRequestComment]).where(
+            PullRequestComment.pull_request_number.in_(numbers)), db)
```

Comment:
But this is exactly what we want to discover, no? Those outliers that ruin a specific metric ("Review Time" in this case)?

Reply:
you're right, that makes sense 👍. How do you handle PR opened long-time ago and not reviewed yet in terms of contribution value to "Review Time"? Is that excluded?

Reply:
Calculating any metric needs an ending timestamp. If it was not approved, we do not take it into account because there is no ending timestamp. This is an interesting and apparently common edge case, I summon @warenlg to the discussion.